### ISSUE TITLE
Asymmetric tiling 

### ISSUE THREAD

**Issue by alejobrainz** [huggingface/diffusers#2633]
Hello.  I'm trying to achieve tiling asymmetrically using Diffusers, in a similar fashion to the asymmetric tiling in Automatic1111's extension https://github.com/tjm35/asymmetric-tiling-sd-webui.

My understanding is that I must traverse all layers to alter the padding, in my case circular in X and constant in Y, but I would love to get advice on how to make a such change to the conv2d system in DIffusers.

Your advice is highly appreciated, as it may also help others down the road facing the same need.

**Comment by patrickvonplaten** [comment]
Hey @alejobrainz,

Sorry I currently won't have time to look into it, but maybe someone in the community is interested!

**Comment by alejobrainz** [comment]
Actually solved it.  Sharing for the community.  Basically I a look at [Shirayu's](https://github.com/shirayu) great [solution ](https://github.com/huggingface/diffusers/issues/556#issuecomment-1253974962) for overriding tiling to circular, and adapted it based on the Asymmetric tiling [extension ](https://github.com/tjm35/asymmetric-tiling-sd-webui/blob/main/scripts/asymmetric_tiling.py#L58-93) from Automatic 1111 WebUI.  

Here is my code for making the result tilable on X only:

```
from typing import Optional
from torch import Tensor
from torch.nn import functional as F
from torch.nn import Conv2d
from torch.nn.modules.utils import _pair

def asymetricConv2DConvForward(self, input: Tensor, weight: Tensor, bias: Optional[Tensor]):
    self.paddingX = (self._reversed_padding_repeated_twice[0], self._reversed_padding_repeated_twice[1], 0, 0)
    self.paddingY = (0, 0, self._reversed_padding_repeated_twice[2], self._reversed_padding_repeated_twice[3])
    working = F.pad(input, self.paddingX, mode='circular')
    working = F.pad(working, self.paddingY, mode='constant')
    return F.conv2d(working, weight, bias, self.stride, _pair(0), self.dilation, self.groups)

targets = [pipe.vae, pipe.text_encoder, pipe.unet,]
conv_layers = []
for target in targets:
    for module in target.modules():
        if isinstance(module, torch.nn.Conv2d):
            conv_layers.append(module)

for cl in conv_layers:
    cl._conv_forward = asymetricConv2DConvForward.__get__(cl, Conv2d)
```

**Comment by alejobrainz** [comment]
Solved.  See above.

**Comment by yijinsheng** [comment]
> 

it works for me in diffusers==0.19.1,  but fails in diffusers==0.19.2
maybe tthe official made some upgrades that caused it failed

**Comment by yijinsheng** [comment]
> Actually solved it. Sharing for the community. Basically I a look at @shirayu's great [solution ](https://github.com/huggingface/diffusers/issues/556#issuecomment-1253974962) for overriding tiling to circular, and adapted it based on the Asymmetric tiling [extension ](https://github.com/tjm35/asymmetric-tiling-sd-webui/blob/main/scripts/asymmetric_tiling.py#L58-93) from Automatic 1111 WebUI.
> 
> Essentially, you collect layers from the vae, text encoder and unet within the pipe, and override their conv2D Forward method, setting a different padding mode for x and y.
> 
> Here is my code for making the result tilable on X only:
> 
> ```
> 
> #...create pipe here
> 
> from typing import Optional
> from torch import Tensor
> from torch.nn import functional as F
> from torch.nn import Conv2d
> from torch.nn.modules.utils import _pair
> 
> def asymmetricConv2DConvForward(self, input: Tensor, weight: Tensor, bias: Optional[Tensor]):
>     self.paddingX = (self._reversed_padding_repeated_twice[0], self._reversed_padding_repeated_twice[1], 0, 0)
>     self.paddingY = (0, 0, self._reversed_padding_repeated_twice[2], self._reversed_padding_repeated_twice[3])
>     working = F.pad(input, self.paddingX, mode='circular')
>     working = F.pad(working, self.paddingY, mode='constant')
>     return F.conv2d(working, weight, bias, self.stride, _pair(0), self.dilation, self.groups)
> 
> targets = [pipe.vae, pipe.text_encoder, pipe.unet,]
> conv_layers = []
> for target in targets:
>     for module in target.modules():
>         if isinstance(module, torch.nn.Conv2d):
>             conv_layers.append(module)
> 
> for cl in conv_layers:
>     cl._conv_forward = asymmetricConv2DConvForward.__get__(cl, Conv2d)
> 
> # ...use pipe here
> ```
it works for me in diffusers==0.19.1,but not working in diffusers==0.19.2
maybe the  official made some upgrades that caused it failed

**Comment by yijinsheng** [comment]
> Yep, this changed with the recent change in diffusers 0.19.2 - [b1e5279#diff-528ec2d3100f053b2b75ba808cb0e8b983545b721b1c42dc1fd47d76522d7faeR130](https://github.com/huggingface/diffusers/commit/b1e52794a28a66b575943fdfeca16809b605cc3c#diff-528ec2d3100f053b2b75ba808cb0e8b983545b721b1c42dc1fd47d76522d7faeR130)
> 
> This workaround works for me. Change the last for-loop to:
> 
> ```python
> for cl in conv_layers:
>     if isinstance(cl, diffusers.models.lora.LoRACompatibleConv) and cl.lora_layer is None:
>         cl.lora_layer = lambda *x: 0
> 
>     cl._conv_forward = asymmetricConv2DConvForward.__get__(cl, torch.nn.Conv2d)
> ```
> 
> This forces it to use the _conv_forward wrapper when a lora isn't set. This workaround should work with loras as well, although I haven't tested it.
yes ,it worked.but i need to set a lora model in my case, so  use 0.19.1 before anyone can solve it ,i won't do an upgrade.

**Comment by cmdr2** [comment]
@yijinsheng The workaround should work with lora as well. Does it fail if you set a lora?

**Comment by yijinsheng** [comment]
> @yijinsheng The workaround should work with lora as well. Does it fail if you set a lora?

sorry late to reply .i have test it and it   work with lora as well,👍

- [ ]

**Comment by cmdr2** [comment]
As a tip, disable VAE Tiling before using this. That is, set `pipe.vae.use_tiling` to `False` before running this code.

**Comment by madaror** [comment]
Hi, if still relevant please take a look on our recent paper, the code can be found in the webpage as well:
https://madaror.github.io/tiled-diffusion.github.io/

**Comment by DaLizardWizard** [comment]
@madaror I have had a look at the paper and code, but cannot seem to get self-tiling along both the X and Y axis working for SD3. Is this something that is supported?
